doc/design/sql_program_dependency_analyze.md

```diff
@@ -0,0 +1,203 @@
+
+
+## Analyzing Database Context
+
+Some SQL program has `USE` clause which defines the database the following SQL statements are using.
+
+```SQL
+USE db;
+CREATE TABLE my_table AS ...
+SELECT * FROM my_table TO TRAIN ...
+USE db2;
+SELECT * FROM my_table_in_db2 TO TRAIN ...
+```
+
+We need to go over the SQL program and get the last `USE` for each SQL statement, and modify the table
+name without prefix `db.` to `db.table` in above step after we get the input/output tables for each
+statement, so that we return the table read/write information with table full name `db.table`.
+
+Note that external parsers do not support `USE` statement, we can do this simply by split the SQL
+program by `;` and using regex to match if the statement is like `USE db_identifier`.
+
+## Hazard
+
+SQL program is just like normal computer programs. We treat tables as variables in our case.
+Analyzing computer programs always have hazards: https://en.wikipedia.org/wiki/Hazard_(computer_architecture). In the example above, we can construct a graph like below:
+
+<p align="center">
+<img src="figures/hazard.png">
+</p>
+
+Note that the last "DROP TABLE" statement must execute after the two "Explain" statements, because
+the table is used by the explain statements before it could be changed (in this case, deleted). This
+case is called "Write After Read". To solve this kind of data hazards, we can add one dependency node
+"table WAR" between the explain statements and drop statement:
+
+<p align="center">
+<img src="figures/hazard_solve.png">
+</p>
+
+We also need to solve "Write After Write" hazard just like "Write After Read" if the SQL program has
+this situation.
+
+## Implementation
+
+In order to construct the dependency graph of the SQL program, we need to analyze the parsed SQL
+statements in every parser, including MySQL parser, Hive parser, calcite parser, and SQLFlow extended
+parser. Since the Hive parser and calcite parser is written in Java, we need to pass the table read/write
+information from Java to Go:
+
+```proto
+// InputOutputTables represents input and output tables a statement manipulates.
+message InputOutputTables {
+   repeated string input_tables = 1;
+   repeated string output_tables = 2;
+}
+
+message ParserResponse {
+  repeated string sql_statements = 1;
+  int32 index = 2;
+  string error = 3;
+  // return tables that each statement manipulates.
+  repeated InputOutputTables input_output_tables = 4;
+}
+```
+
+When parsing the SQL program in Go, we can then get table read/write information for each standard
+SQL statement, then parse the extended SQL to get model read/write information, then construct the
+graph:
+
+```go
+type SQLProgramNode interface {
+   IsStatementNode() bool
+   GetInputs() *[]Node
+   GetOutputs() *[]Node
+}
+
+type StatementNode struct {
+   Statement string
+   DockerImage string
+   // StatementNode's input/output must be a table.
+   Inputs *[]TableNode
+   Outputs *[]TableNode
+}
+
+type TableNode struct {
```

Comment:
TableNode => ir.Table or parser.Table ?

Reply:
Updated, we define these structs under package `ir`. I keep the name `TableNode` because the name `ir.Table` should represent the parse result of a table name or table alias.